fs2-google-pubsub-grpc/src/main/scala/com/permutive/pubsub/consumer/grpc/internal/PubsubSubscriber.scala

```diff
@@ -53,31 +49,43 @@ private[consumer] object PubsubSubscriber {
             .getOrElse(builder)
             .build()
 
-        sub.addListener(new PubsubErrorListener(messages), MoreExecutors.directExecutor)
+        sub.addListener(new PubsubErrorListener(queue), MoreExecutors.directExecutor)
 
-        val service = sub.startAsync()
-        val shutdown =
-          F.blocking(
-            service.stopAsync().awaitTerminated(config.awaitTerminatePeriod.toSeconds, TimeUnit.SECONDS)
-          ).handleErrorWith(config.onFailedTerminate)
-
-        (messages, shutdown)
+        sub.startAsync()
       }
-    }
+    )(service =>
+      Sync[F]
+        .blocking(service.stopAsync().awaitTerminated(config.awaitTerminatePeriod.toSeconds, TimeUnit.SECONDS))
+        .handleErrorWith(config.onFailedTerminate)
+    )
 
-  class PubsubErrorListener[R](messages: BlockingQueue[Either[InternalPubSubError, R]]) extends ApiService.Listener {
+  class PubsubMessageReceiver[F[_]: Sync, L](queue: BlockingQueue[Either[L, Model.Record[F]]]) extends MessageReceiver {
+    override def receiveMessage(message: PubsubMessage, consumer: AckReplyConsumer): Unit =
+      queue.put(Right(Model.Record(message, Sync[F].delay(consumer.ack()), Sync[F].delay(consumer.nack()))))
+  }
+
+  class PubsubErrorListener[R](queue: BlockingQueue[Either[InternalPubSubError, R]]) extends ApiService.Listener {
     override def failed(from: ApiService.State, failure: Throwable): Unit =
-      messages.put(Left(InternalPubSubError(failure)))
+      queue.put(Left(InternalPubSubError(failure)))
   }
 
+  def takeNextElement[F[_]: Sync, A](messages: BlockingQueue[A]): F[A] =
+    for {
+      nextOpt <- Sync[F].delay(Option(messages.poll()))
+      next    <- nextOpt.fold(Sync[F].blocking(messages.take()))(Applicative[F].pure)
+    } yield next
+
```

Comment:
This is the key change.

Reply:
Was curious to see if `LinkedBlockingQueue#poll()` was blocking, so check the implementation and looks good to me:
```java
    public E poll() {
        final AtomicInteger count = this.count;
        if (count.get() == 0)
            return null;
```

Reply:
Yeah, the docstring says it was non-blocking. Previously we only used `take`. I'll add a comment to make this clear :+1:

Reply:
Ah sorry, the docstring does not mention not blocking. I had done the same check of the code myself to be sure!